### PR TITLE
Add create vnic details

### PR DESCRIPTION
**Pull request by joekr**
**What this PR does / why we need it**:
Add support for Windows. Windows on OCI requires Bare Metal currently. For Hyper-V to work on Bare Metal the second nic has to work, hence adding support for the second vnic.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

**Comment by joekr** [comment]
- [ ] Squash and rebase commits before merging.

File: cloud/scope/vnic_reconciler.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+		vnicId, err := m.createVnicAttachment(ctx, vnicAttachment)
+		if err != nil {
+			msg := fmt.Sprintf("Error creating VnicAttachment %s for cluster %s",
+				*vnicAttachment.DisplayName, m.Cluster.Name)
+			m.Logger.Error(err, msg)
+			return err
+		}
+		m.OCIMachine.Spec.VnicAttachments[index].VnicAttachmentId = vnicId
+		vnicAttachment.VnicAttachmentId = common.String("TESTING")
```
Comment (joekr) [inline cloud/scope/vnic_reconciler.go:34]: I need to fix this

File: api/v1beta1/types.go [hunk 2]
```diff
@@ -64,6 +64,30 @@ type NetworkDetails struct {
 	AssignPrivateDnsRecord *bool `json:"assignPrivateDnsRecord,omitempty"`
 }
 
+type VnicAttachment struct {
+	// VnicAttachmentId defines the ID of the VnicAttachment
+	VnicAttachmentId *string `json:"vnicAttachmentId,omitempty"`
+
+	// AssignPublicIp defines whether the instance should have a public IP address
```
Comment (shyamradhakrishnan) [inline api/v1beta1/types.go:71]: whether the instance -> whether the vnic?

File: cloud/scope/vnic_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		subnetId, err = m.getMachineSubnet(spec.SubnetName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tags := m.getFreeFormTags(*m.OCICluster)
+
+	definedTags := make(map[string]map[string]interface{})
+	if m.OCIMachine.Spec.DefinedTags != nil {
```
Comment (shyamradhakrishnan) [inline cloud/scope/vnic_reconciler.go:56]: can we use a common function for this at machine.go? or do we have one?

File: cloud/scope/vnic_reconciler.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+			FreeformTags:           tags,
+			DefinedTags:            definedTags,
+			HostnameLabel:          m.OCIMachine.Spec.NetworkDetails.HostnameLabel,
+			SkipSourceDestCheck:    m.OCIMachine.Spec.NetworkDetails.SkipSourceDestCheck,
+			AssignPrivateDnsRecord: m.OCIMachine.Spec.NetworkDetails.AssignPrivateDnsRecord,
+			DisplayName:            vnicName,
+		},
+	}
+
+	retryToken := fmt.Sprintf("%s-%s", string(m.OCIMachine.UID), *vnicName)
```
Comment (shyamradhakrishnan) [inline cloud/scope/vnic_reconciler.go:86]: do we need the retry token here? I would suggest we remove this a,d put it back if we his issues, especially with the character limit issues

File: cloud/scope/vnic_reconciler.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	req := core.AttachVnicRequest{AttachVnicDetails: secondVnic,
+		OpcRetryToken: ociutil.GetOPCRetryToken(retryToken)}
+	resp, err := m.ComputeClient.AttachVnic(ctx, req)
+	m.Logger.Info("AttachVnic resp: ", resp)
```
Comment (shyamradhakrishnan) [inline cloud/scope/vnic_reconciler.go:90]: please remove this line

File: cloud/scope/vnic_reconciler.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.VnicId, nil
+}
+
+func (m *MachineScope) updateVnicAttachment(ctx context.Context, spec infrastructurev1beta1.VnicAttachment) error {
```
Comment (shyamradhakrishnan) [inline cloud/scope/vnic_reconciler.go:99]: where are we calling this?

File: cloud/scope/vnic_reconciler.go [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+	"github.com/pkg/errors"
+)
+
+func (m *MachineScope) ReconcileVnicAttachments(ctx context.Context) error {
+	if m.IsControlPlane() {
+		return errors.New("cannot attach multiple vnics to ControlPlane machines")
+	}
+
+	for index, vnicAttachment := range m.OCIMachine.Spec.VnicAttachments {
+		if m.vnicAttachmentExists(ctx, vnicAttachment) {
```
Comment (shyamradhakrishnan) [inline cloud/scope/vnic_reconciler.go:20]: this will be an infinite loop if we dont do comparison right?

File: cloud/scope/vnic_reconciler.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
+package scope
```
Comment (shyamradhakrishnan) [inline cloud/scope/vnic_reconciler.go:1]: so that I keep track , please add a unit test file

File: controllers/ocimachine_controller.go [hunk 9]
```diff
@@ -289,6 +289,17 @@ func (r *OCIMachineReconciler) reconcileNormal(ctx context.Context, logger logr.
...
 		}
+
+		if len(machine.Spec.VnicAttachments) > 0 {
+			err := machineScope.ReconcileVnicAttachments(ctx)
+			if err != nil {
+				r.Recorder.Event(machine, corev1.EventTypeWarning, "ReconcileError", errors.Wrapf(err, "failed to reconcile OCIMachine").Error())
+				conditions.MarkFalse(machineScope.OCIMachine, infrastructurev1beta1.InstanceReadyCondition, infrastructurev1beta1.InstanceVnicAttachmentFailedReason, clusterv1.ConditionSeverityError, "")
+				return ctrl.Result{}, err
+			}
+			machineScope.Info("Instance vnic attachment success")
```
Comment (shyamradhakrishnan) [inline controllers/ocimachine_controller.go:300]: should we add ad codition true here with vnic attachement succeeded?

File: test/e2e/cluster_test.go [hunk 10]
```diff
@@ -199,6 +199,29 @@ var _ = Describe("Workload cluster creation", func() {
 		validateOLImage(namespace.Name, clusterName)
 	})
 
+	It("Windows - With 1 Linux control-plane nodes and with 1 Windows worker nodes", func() {
```
Comment (shyamradhakrishnan) [inline test/e2e/cluster_test.go:202]: better to add calico in unit test name

File: templates/cluster-template-windows-calico.yaml [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+          kubeletExtraArgs:
+            cloud-provider: external
+            provider-id: oci://{{ ds.meta_data["instance_id"] }}
+            feature-gates: WindowsHostProcessContainers=true
+            v: "2"
+            windows-priorityclass: ABOVE_NORMAL_PRIORITY_CLASS
+          name: '{{ ds.meta_data["local_hostname"] }}'
+      preKubeadmCommands:
+        - powershell C:\Windows\Setup\Scripts\enable_second_nic.ps1
+        - powershell C:\Users\opc\attach_secondary_vnic.ps1 > C:\Users\opc\attach_secondary_vnic_log.txt
```
Comment (shyamradhakrishnan) [inline templates/cluster-template-windows-calico.yaml:133]: if you have a windows doc ticket, please add these points which have to explained in the doc do that wehen we review the doc, we dont miss explaining to the user what all should be done in the templates

File: templates/cluster-template-windows-calico.yaml [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+  replicas: ${NODE_MACHINE_COUNT}
+  selector:
+    matchLabels:
+  template:
+    spec:
+      clusterName: "${CLUSTER_NAME}"
+      version: "${KUBERNETES_VERSION}"
+      bootstrap:
+        configRef:
+          name: "${CLUSTER_NAME}-md-win"
```
Comment (shyamradhakrishnan) [inline templates/cluster-template-windows-calico.yaml:150]: md-windows?

File: templates/cluster-template-windows-calico.yaml [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+          apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
+          kind: KubeadmConfigTemplate
+      infrastructureRef:
+        name: "${CLUSTER_NAME}-md-win"
```
Comment (shyamradhakrishnan) [inline templates/cluster-template-windows-calico.yaml:154]: md-windows?

File: templates/cluster-template-windows-calico.yaml (hunk 12 above)
Comment (joekr) [reply templates/cluster-template-windows-calico.yaml:150]: We are limited in the cluster name size.

File: templates/cluster-template-windows-calico.yaml (hunk 13 above)
Comment (joekr) [reply templates/cluster-template-windows-calico.yaml:154]: We are limited in the cluster name size.

File: api/v1beta1/types.go (hunk 2 above)
Comment (joekr) [reply api/v1beta1/types.go:71]: fixed

File: cloud/scope/vnic_reconciler.go (hunk 3 above)
Comment (joekr) [reply cloud/scope/vnic_reconciler.go:56]: fixed

File: cloud/scope/vnic_reconciler.go (hunk 4 above)
Comment (joekr) [reply cloud/scope/vnic_reconciler.go:86]: updated

File: cloud/scope/vnic_reconciler.go (hunk 5 above)
Comment (joekr) [reply cloud/scope/vnic_reconciler.go:90]: updated

File: cloud/scope/vnic_reconciler.go (hunk 7 above)
Comment (joekr) [reply cloud/scope/vnic_reconciler.go:20]: updated

File: controllers/ocimachine_controller.go (hunk 9 above)
Comment (joekr) [reply controllers/ocimachine_controller.go:300]: great idea

File: templates/cluster-template-windows-calico.yaml (hunk 11 above)
Comment (joekr) [reply templates/cluster-template-windows-calico.yaml:133]: good call will add that ticket and call this out.

Comment (joekr) [reply templates/cluster-template-windows-calico.yaml:133]: circular dependency link https://github.com/oracle/cluster-api-provider-oci/issues/206

File: cloud/scope/vnic_reconciler_test.go [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+								DisplayName: common.String("vnicDisplayName"),
+							},
+							{
+								InstanceId:  common.String("test"),
+								DisplayName: common.String("vnicDisplayName"),
+							},
+						},
+					}, nil)
+
+				// TODO: should work but isn't currently
```
Comment (shyamradhakrishnan) [inline cloud/scope/vnic_reconciler_test.go:121]: adding comment so that we remove this before merging

**Comment by joekr** [comment]
# unit tests

```
?       github.com/oracle/cluster-api-provider-oci      [no test files]
ok      github.com/oracle/cluster-api-provider-oci/api/v1beta1  0.275s  coverage: 17.6% of statements
?       github.com/oracle/cluster-api-provider-oci/cloud/config [no test files]
ok      github.com/oracle/cluster-api-provider-oci/cloud/ociutil        0.546s  coverage: 25.0% of statements
ok      github.com/oracle/cluster-api-provider-oci/cloud/scope  166.030s        coverage: 76.5% of statements
?       github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks    [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/base  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/base/mock_base        [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute       [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement     [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement/mock_computemanagement      [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine       [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine/mock_containerengine  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity      [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity        [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb  [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn   [no test files]
?       github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn  [no test files]
ok      github.com/oracle/cluster-api-provider-oci/controllers  26.702s coverage: 67.7% of statements
ok      github.com/oracle/cluster-api-provider-oci/exp/api/v1beta1      0.475s  coverage: 8.8% of statements
ok      github.com/oracle/cluster-api-provider-oci/exp/controllers      1.353s  coverage: 55.3% of statements
?       github.com/oracle/cluster-api-provider-oci/feature      [no test files]
?       github.com/oracle/cluster-api-provider-oci/version      [no test files]
```

# e2e tests

File: scripts/ci-e2e.sh [hunk 15]
```diff
@@ -23,6 +23,7 @@ source "${REPO_ROOT}/hack/ensure-tags.sh"
 : "${OCI_ORACLE_LINUX_IMAGE_ID:?Environment variable empty or not defined.}"
 : "${OCI_UPGRADE_IMAGE_ID:?Environment variable empty or not defined.}"
 : "${OCI_ALTERNATIVE_REGION_IMAGE_ID:?Environment variable empty or not defined.}"
+: "${OCI_WINDOWS_IMAGE_ID:?Environment variable empty or not defined.}"
```
Comment (joekr) [inline scripts/ci-e2e.sh:26]: @shyamradhakrishnan want to specifically point out this new requirement in order to run e2e tests.

Comment (shyamradhakrishnan) [reply scripts/ci-e2e.sh:27]: I think we should keep it optoional right?

File: cloud/scope/vnic_reconciler_test.go [hunk 16]
```diff
@@ -0,0 +1,197 @@
+package scope
```
Comment (shyamradhakrishnan) [inline cloud/scope/vnic_reconciler_test.go:1]: license?

File: cloud/scope/vnic_reconciler.go [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
```
Comment (shyamradhakrishnan) [inline cloud/scope/vnic_reconciler.go:1]: license?

Comment (joekr) [reply cloud/scope/vnic_reconciler.go:17]: updated

**Review by shyamradhakrishnan — APPROVED** [review]

**Comment by joekr** [comment]
After removing the windows template and e2e tests I reran the tests all passed except `managed cluster` but I think that is a setup issue on my end.

**Review by shyamradhakrishnan — APPROVED** [review]

**Review by shyamradhakrishnan — APPROVED** [review]

**Pull request merged**